### PR TITLE
add oc set volume support for configmaps

### PR DESCRIPTION
**Pull request by deads2k**
`oc set volume` didn't support configmaps.  This adds that feature.

It builds on the previous pull, so last commit only.

@fabianofranz

### CODE REVIEW

**Comment by deads2k** [comment]
@stevekuznetsov take a look too.

File: pkg/cmd/cli/cmd/set/volume.go [hunk 1]
```diff
@@ -249,6 +251,10 @@ func (a *AddVolumeOptions) Validate(isAddOp bool) error {
 			a.Type = "secret"
 			a.TypeChanged = true
 		}
+		if len(a.Type) == 0 && (len(a.ConfigMapName) > 0) {
```
Comment (stevekuznetsov) [inline pkg/cmd/cli/cmd/set/volume.go]: It isn't entirely certain to me why if this is an add operation and a type and config map name are both provided, that the type should prevail.

File: test/cmd/volumes.sh [hunk 2]
```diff
@@ -29,6 +29,7 @@ os::cmd::expect_success 'oc volume dc/test-deployment-config --add --name=vol0 -
 os::cmd::expect_success 'oc set volume dc/test-deployment-config --add --name=vol2 --type=emptydir -m /opt'
 os::cmd::expect_failure_and_text "oc set volume dc/test-deployment-config --add --name=vol1 --type=secret --secret-name='\$ecret' -m /data" 'overwrite to replace'
 os::cmd::expect_success "oc set volume dc/test-deployment-config --add --name=vol10 --secret-name='my-secret' -m /data-2"
+os::cmd::expect_success "oc set volume dc/test-deployment-config --add --name=vol11 --configmap-name='my-configmap' -m /data-21"
```
Comment (stevekuznetsov) [inline test/cmd/volumes.sh]: Don't we want to probe the resulting object with a `jsonpath` to ensure that what we did worked? For all of the tests in that vein, as well. I don't really see the current test as being that useful, `os.Exit(0)` would pass it.

File: pkg/cmd/cli/cmd/set/volume.go [hunk 3]
```diff
@@ -274,6 +280,10 @@ func (a *AddVolumeOptions) Validate(isAddOp bool) error {
 				if len(a.SecretName) == 0 {
 					return errors.New("must provide --secret-name for --type=secret")
 				}
+			case "configmap":
```
Comment (stevekuznetsov) [inline pkg/cmd/cli/cmd/set/volume.go]: I have a feeling you'll prefer to defer this for a follow-up, but this validation logic is *really* hard to make sense of.

File: pkg/cmd/cli/cmd/set/volume.go (hunk 1 above)
Comment (deads2k) [inline pkg/cmd/cli/cmd/set/volume.go]: > It isn't entirely certain to me why if this is an add operation and a type and config map name are both provided, that the type should prevail.

typos.  If I specify conflicting values, I should fail either before or after I get here.  Now I don't actually see that logic ( :( ), but that's the reasoning behind this and the existing code.

File: pkg/cmd/cli/cmd/set/volume.go [hunk 4]
```diff
@@ -274,6 +280,10 @@ func (a *AddVolumeOptions) Validate(isAddOp bool) error {
...
+				if len(a.ConfigMapName) == 0 {
+					return errors.New("must provide --configmap-name for --type=configmap")
```
Comment (stevekuznetsov) [inline pkg/cmd/cli/cmd/set/volume.go]: @fabianofranz do we not have a better mechanism to ensure that this relationship is enforced?

File: pkg/cmd/cli/cmd/set/volume.go (hunk 3 above)
Comment (deads2k) [inline pkg/cmd/cli/cmd/set/volume.go]: > I have a feeling you'll prefer to defer this for a follow-up, but this validation logic is really hard to make sense of.

Yeah, tightening validation and fixing the method overall is a TODO kind of item.

File: test/cmd/volumes.sh (hunk 2 above)
Comment (deads2k) [inline test/cmd/volumes.sh]: > Don't we want to probe the resulting object with a jsonpath to ensure that what we did worked? For all of the tests in that vein, as well. I don't really see the current test as being that useful, os.Exit(0) would pass it.

Added.

**Comment by stevekuznetsov** [comment]
LGTM from me, would like to know if there's a better way to describe the "one flag is set so another is required" paradigm than what we have, though.

**Comment by deads2k** [comment]
[merge]

> LGTM from me, would like to know if there's a better way to describe the "one flag is set so another is required" paradigm than what we have, though.

I'm not aware of anything better than checking against a known good set.

**Pull request merged**